src/wp-includes/global-styles-and-settings.php

```diff
@@ -83,25 +83,33 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
  * @return string Stylesheet.
  */
 function wp_get_global_stylesheet( $types = array() ) {
-	// Return cached value if it can be used and exists.
-	// It's cached by theme to make sure that theme switching clears the cache.
-	$can_use_cached = (
-		( empty( $types ) ) &&
-		( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) &&
-		( ! defined( 'SCRIPT_DEBUG' ) || ! SCRIPT_DEBUG ) &&
-		( ! defined( 'REST_REQUEST' ) || ! REST_REQUEST ) &&
-		! is_admin()
-	);
-	$transient_name = 'global_styles_' . get_stylesheet();
+	// Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme developers workflow.
+	$can_use_cached = empty( $types ) && ! WP_DEBUG;
+	/*
+	 * By using the 'theme_json' group, this data is marked to be non-persistent across requests.
+	 * See `wp_cache_add_non_persistent_groups` in src/wp-includes/load.php and other places.
+	 *
+	 * The rationale for this is to make sure derived data from theme.json
+	 * is always fresh from the potential modifications done via hooks
+	 * that can use dynamic data (modify the stylesheet depending on some option,
+	 * settings depending on user permissions, etc.).
+	 * See some of the existing hooks to modify theme.json behaviour:
+	 * https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/
+	 *
+	 * A different alternative considered was to invalidate the cache upon certain
+	 * events such as options add/update/delete, user meta, etc.
+	 * It was judged not enough, hence this approach.
+	 * See https://github.com/WordPress/gutenberg/pull/45372
+	 */
+	$cache_group    = 'theme_json';
+	$cache_key      = 'wp_get_global_stylesheet';
```

Comment:
Eventually it would be nice to also have different cache keys depending on `$types` (i.e. when it's not empty), but this can be a separate follow up. How common is it to call this function with `$types` being non-empty?

Reply:
In the front-end, we only call it with no parameters ([code](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/script-loader.php#L2434)).

The other cases are for the editor, where we need more granularity in how we enqueue the stylesheet ([here](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/script-loader.php#L2953), and [here](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/block-editor.php#L413)) as the editor canvas (content) and the editor sidebar (inspector controls) need different pieces of information and have different CSS scopes.

Reply:
It would be nice to have the current theme stylesheet in this cache key too, so we don't have to do the same cache invalidation.